packages/SwingSet/src/types-external.js

```diff
@@ -366,4 +367,45 @@ export {};
  *
  * @typedef { { enableDisavow?: boolean } } HasEnableDisavow
  * @typedef { DynamicVatOptions & HasEnableDisavow } StaticVatOptions
+ *
+ * @typedef { { vatParameters?: *, upgradeMessage: string } } VatUpgradeOptions
+ * @typedef { { incarnationNumber: number } } VatUpgradeResults
+ *
+ * @callback ShutdownWithFailure
+ * Called to shut something down because something went wrong, where the reason
+ * is supposed to be an Error that describes what went wrong. Some valid
+ * implementations of `ShutdownWithFailure` will never return, either
+ * because they throw or because they immediately shutdown the enclosing unit
+ * of computation. However, they also might return, so the caller should
+ * follow this call by their own defensive `throw reason;` if appropriate.
+ *
+ * @param {Error} reason
+ * @returns {void}
+ *
+ * @typedef {object} VatAdminFacet
+ * A powerful object that can be used to terminate the vat in which a
+ * contract is running, upgrade it with new code or parameters, get
+ * statistics, or be notified when it terminates.
```

Comment:
Is it appropriate to mention "contract" here? I thought vatAdmin supports manipulation of _arbitrary_ vats.
```suggestion
 * @typedef {object} VatAdminFacet
 * A powerful object corresponding with a vat
 * that can be used to upgrade it with new code or parameters,
 * terminate it, or be notified when it terminates.
```

Reply:
good catch, the types were written from the POV of a zoe user, but yes they're definitely not contract-specific